payments/stripe/test_stripe.py

```diff
@@ -95,8 +96,8 @@ def test_form_contains_stripe_script(self):
         form = provider.get_form(payment)
         self.assertTrue(
             '<script class="stripe-button" data-amount="10000" '
-            'data-currency="USD" data-description="payment" data-image="" '
-            'data-key="%s" data-name="%s" '
+            'data-currency="USD" data-description="payment" data-email="john@doe.com" '
```

Comment:
You should test with `payment.billing_email = None`.

You keep this assert as is, it's fine. Then, below it, you do like so:
```python
payment.billing_email = None
form = provider.get_form(payment)
self.assertTrue(...)
```

Reply:
Sure, this is not a problem :)

What I was trying to validate is the `stripe.Charge` creation with the valid `receipt_email`. I tried with no luck to patch the payment.save and the payment.is_valid, something like:

```python

@contextmanager
def mock_stripe_Payment_save():
    with patch('payments.stripe.forms.ModalPaymentForm.save') as mocked_payment_save:
        ...
        ...
        yield mocked_payment_save

@contextmanager
def mock_stripe_get_form():
    with patch('payments.stripe.forms.ModalPaymentForm.is_valid') as mocked_stripe_get_form:
        ....
        ...
        yield mocked_stripe_get_form

...
...

class TestStripeProvider(TestCase):

    ...
    ...

    def test_provider_charge_contains_receipt_email(self):
        payment = Payment()
        provider = StripeProvider(
            name='Example.com store',
            secret_key=SECRET_KEY, public_key=PUBLIC_KEY)
        data = {'stripeToken': 'abcd'}

        with mock_stripe_Charge_create(error_msg=None):
            with mock_stripe_Charge_retrieve():
                with mock_stripe_Payment_save():
                    with mock_stripe_get_form():
                        form = provider.get_form(payment, data=data)
                        self.assertTrue(
                            form.charge
                            and "receipt_email" in form.charge
                            and form.charge.__dict__.get('_previous', {}).get('receipt_email', False) == "john@doe.com"
                        )
```

also explored the way to create manually the form, inject the cleaned_data and request the 

```python
                form = ModalPaymentForm(
                    data=data, payment=payment, provider=provider
                )
```
, but I've no more time to investigate... :(

Reply:
No need to test that. It's Stripe's job at this level.